### PR TITLE
enh(ldap): improve performance of ldap sync on login

### PR DESCRIPTION
**Pull request by kduret**
## Description

improve performance of ldap sync on login

**Fixes** MON-12076

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [x] 20.10.x
- [x] 21.04.x
- [x] 21.10.x
- [x] 22.04.x (master)

### CODE REVIEW

File: www/class/centreonContactgroup.class.php [hunk 1]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
         return $cg_name . " (LDAP : " . $ldap_name . ")";
     }
 
+    /**
+     * find LDAP group id by name
+     *
+     * @param int $ldapId
+     * @param string $name
+     * @return int
```
Comment (kduret) [inline www/class/centreonContactgroup.class.php:159]:
```suggestion
     * @return int|null
```

File: www/class/centreonContactgroup.class.php [hunk 2]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+
+        return $ldapGroupId;
+    }
+
+    /**
+     * find LDAP group id by dn
+     *
+     * @param int $ldapId
+     * @param string $dn
+     * @return int
```
Comment (kduret) [inline www/class/centreonContactgroup.class.php:186]:
```suggestion
     * @return int|null
```

File: www/class/centreonContactgroup.class.php [hunk 3]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+        return $ldapGroupId;
+    }
+
+    /**
+     * Insert ldap group if does not exist
+     *
+     * @param int $ldapId
+     * @param string $name
+     * @param string $dn
+     * @return int
```
Comment (kduret) [inline www/class/centreonContactgroup.class.php:214]:
```suggestion
     * @return int|null
```

File: www/class/centreonContactgroup.class.php [hunk 4]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
     }
 
+    /**
+     * find LDAP group id by name
+     *
+     * @param int $ldapId
+     * @param string $name
+     * @return int|null
+     */
+    private function findLdapGroupIdByName($ldapId, $name)
```
Comment (jeremyjaouen) [inline www/class/centreonContactgroup.class.php:161]: typehint parameters

File: www/class/centreonContactgroup.class.php [hunk 5]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+    }
+
+    /**
+     * find LDAP group id by dn
+     *
+     * @param int $ldapId
+     * @param string $dn
+     * @return int|null
+     */
+    private function findLdapGroupIdByDn($ldapId, $dn)
```
Comment (jeremyjaouen) [inline www/class/centreonContactgroup.class.php:188]: typehint parameters

File: www/class/centreonContactgroup.class.php [hunk 6]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+
+    /**
+     * Insert ldap group if does not exist
+     *
+     * @param int $ldapId
+     * @param string $name
+     * @param string $dn
+     * @return int|null
+     */
+    public function insertLdapGroupByNameAndDn($ldapId, $name, $dn)
```
Comment (jeremyjaouen) [inline www/class/centreonContactgroup.class.php:216]: typehint parameters

File: www/class/centreonContactgroup.class.php [hunk 7]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
     /**
      * Insert the ldap groups in table contactgroups
      *
-     * @param string $cg_name : The ldap group name
-     * @return int $row['cg_id'] :The contactgroup id or 0 if error
+     * @param string $cgName The ldap group name
+     * @return int The contactgroup id or null if not found
      */
-    public function insertLdapGroup($cg_name)
+    public function insertLdapGroup($cgName)
```
Comment (jeremyjaouen) [inline www/class/centreonContactgroup.class.php:241]: typehint parameters (if you want as the methods was already existing)

File: www/class/centreonContactgroup.class.php [hunk 8]
```diff
@@ -324,11 +394,16 @@ public function syncWithLdap()
                 foreach ($registeredGroupsFromDB as $registeredGroupFromDB) {
                     $registeredGroups[] = $registeredGroupFromDB['cg_name'];
                 }
+
+                $time = microtime(true);
                 $ldapGroups = $ldapConn->listOfGroups();
-                $toInsertGroups = array_diff($ldapGroups, $registeredGroups);
+                var_dump($ldapGroups);
+                var_dump(microtime(true) - $time);
```
Comment (jeremyjaouen) [inline www/class/centreonContactgroup.class.php:401]: remove var_dump

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

File: www/class/centreonContactgroup.class.php [hunk 9]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+            . "FROM contactgroup "
+            . "WHERE cg_name = :name "
+            . "AND ar_id = :ldapId";
+        $statement = $this->db->prepare($query);
+        $statement->bindValue(':name', $name, \PDO::PARAM_STR);
+        $statement->bindValue(':ldapId', $ldapId, \PDO::PARAM_INT);
+        $statement->execute();
+
+        if ($row = $statement->fetch()) {
+            $ldapGroupId = $row['cg_id'];
```
Comment (callapa) [inline www/class/centreonContactgroup.class.php:175]:
```suggestion
            $ldapGroupId = (int) $row['cg_id'];
```

File: www/class/centreonContactgroup.class.php [hunk 10]
```diff
@@ -149,55 +151,123 @@ public function formatLdapContactgroupName($cg_name, $ldap_name)
...
+            . "FROM contactgroup "
+            . "WHERE cg_ldap_dn = :dn "
+            . "AND ar_id = :ldapId";
+        $statement = $this->db->prepare($query);
+        $statement->bindValue(':dn', $dn, \PDO::PARAM_STR);
+        $statement->bindValue(':ldapId', $ldapId, \PDO::PARAM_INT);
+        $statement->execute();
+
+        if ($row = $statement->fetch()) {
+            $ldapGroupId = $row['cg_id'];
```
Comment (callapa) [inline www/class/centreonContactgroup.class.php:202]:
```suggestion
            $ldapGroupId = (int) $row['cg_id'];
```

**Review by callapa — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

**Pull request merged**